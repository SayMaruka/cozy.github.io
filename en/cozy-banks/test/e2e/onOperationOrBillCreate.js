process.env.NODE_ENV = 'development'

const fs = require('fs')
const path = require('path')
const {
  credentialsFromACHTokenFile,
  toMatchSnapshot,
  spawn,
  prompt,
  couch
} = require('./utils')
const log = require('cozy-logger').namespace('e2e-onOperationOrBillCreate')

const PREFIX = 'cozy-banks-e2e-onOperationOrBillCreate'
const TOKEN_FILE = `/tmp/${PREFIX}-token.json`

/**
 * Spawns ACH, adds token argument automatically
 */
const ach = args => {
  args = args.slice()
  args.splice(1, 0, '-t', TOKEN_FILE)
  log('debug', JSON.stringify(args))
  return spawn('ACH', args)
}

const makeToken = async () => {
  log('info', 'Making token...')
  await spawn('bash', [
    '-c',
    `scripts/make-token cozy.tools:8080 | jq -R '{token: .}' > ${TOKEN_FILE}`
  ])
}

const dropData = async () => {
  log('info', 'Dropping data...')
  const doctypes = [
    'io.cozy.bank.operations',
    'io.cozy.bills',
    'io.cozy.bank.settings',
    'io.cozy.bank.accounts'
  ]
  return couch.dropDatabases(doctypes)
}

const loadData = async () => {
  log('info', 'Loading data...')
  const dir = 'test/fixtures/matching-service'
  for (let fixture of fs.readdirSync(dir)) {
    await ach(['import', path.join(dir, fixture)])
  }
}

const exportAndSnapshot = async () => {
  log('info', 'Exporting and snapshotting...')
  const exportFilename = `/tmp/${PREFIX}-export.json`
  await ach(['export', 'io.cozy.bank.operations,io.cozy.bills', exportFilename])
  const actual = fs.readFileSync(exportFilename).toString()
  const testTitle = 'onOperationOrBillCreate'
  const snapResult = toMatchSnapshot(actual, __filename, testTitle)
  if (snapResult.pass) {
    log('info', 'Snapshot OK !')
  } else {
    console.error(snapResult.report()) // eslint-disable-line no-console
    throw new Error('Snapshot not OK')
  }
}

const runService = async options => {
  log('info', 'Running service...')
  process.env.COZY_URL = 'http://cozy.tools:8080'
  process.env.COZY_CREDENTIALS = JSON.stringify(
    credentialsFromACHTokenFile(TOKEN_FILE)
  )
  await spawn('node', [
    'build/onOperationOrBillCreate.js',
    JSON.stringify(options)
  ])
}

const testService = async options => {
  await dropData()
  await loadData()
  await runService(options)
  await exportAndSnapshot()
}

const main = async () => {
  const e = await prompt('Clear data and run e2e test (y) ?')
  if (e !== 'y') {
    log('info', 'Aborting...')
    return
  }
  await makeToken()
  await testService({ transactionMatching: false })
  await testService({ billMatching: false })
}

main().catch(e => {
  console.error(e) // eslint-disable-line no-console
  process.exit(1)
})
