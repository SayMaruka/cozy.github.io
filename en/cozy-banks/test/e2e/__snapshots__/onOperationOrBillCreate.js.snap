// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`onOperationOrBillCreate 1`] = `
"{
  \\"io.cozy.bank.operations\\": [
    {
      \\"_id\\": \\"ikea_operation\\",
      \\"account\\": \\"compteisa1\\",
      \\"amount\\": -15,
      \\"automaticCategoryId\\": \\"401060\\",
      \\"currency\\": \\"EUR\\",
      \\"date\\": \\"2018-10-17T00:00:00+02:00\\",
      \\"label\\": \\"IKEA\\",
      \\"originalBankLabel\\": \\"IKEA\\",
      \\"type\\": \\"credit card\\",
      \\"bills\\": [
        \\"io.cozy.bills:ikea_bill\\"
      ]
    },
    {
      \\"_id\\": \\"remboursement_cpam\\",
      \\"account\\": \\"compteisa1\\",
      \\"amount\\": 5,
      \\"automaticCategoryId\\": \\"400610\\",
      \\"currency\\": \\"EUR\\",
      \\"date\\": \\"2018-10-18T00:00:00+02:00\\",
      \\"label\\": \\"Remboursement CPAM\\",
      \\"bills\\": [
        \\"io.cozy.bills:ameli_bill\\"
      ]
    },
    {
      \\"_id\\": \\"visite_medecin\\",
      \\"account\\": \\"compteisa1\\",
      \\"amount\\": -20,
      \\"automaticCategoryId\\": \\"400610\\",
      \\"currency\\": \\"EUR\\",
      \\"date\\": \\"2018-10-18T00:00:00+02:00\\",
      \\"label\\": \\"Visite chez le m√©decin\\",
      \\"bills\\": [
        \\"io.cozy.bills:ameli_bill\\"
      ],
      \\"reimbursements\\": [
        {
          \\"billId\\": \\"io.cozy.bills:ameli_bill\\",
          \\"amount\\": 5,
          \\"operationId\\": \\"remboursement_cpam\\"
        }
      ]
    }
  ],
  \\"io.cozy.bills\\": [
    {
      \\"_id\\": \\"ameli_bill\\",
      \\"amount\\": 5,
      \\"currency\\": \\"EUR\\",
      \\"date\\": \\"2018-10-18T00:00:00.000Z\\",
      \\"invoice\\": \\"io.cozy.files:ameli_invoice\\",
      \\"isRefund\\": true,
      \\"originalAmount\\": 20,
      \\"originalDate\\": \\"2018-10-18T00:00:00.000Z\\",
      \\"type\\": \\"health_costs\\",
      \\"vendor\\": \\"Ameli\\"
    },
    {
      \\"_id\\": \\"ikea_bill\\",
      \\"amount\\": 15,
      \\"date\\": \\"2018-10-17T00:00:00Z\\",
      \\"invoice\\": \\"io.cozy.files:ikea_invoice_id\\",
      \\"vendor\\": \\"Ikea\\"
    }
  ]
}"
`;
